oxidized-cuda-kernels/src/nonzero.rs

```diff
@@ -0,0 +1,114 @@
+    }
+
+    let n_elem = layout.shape().elem_count();
+    let n_dim = layout.shape().dims().len();
+    let shape = layout.shape().dims();
+    let stream = storage.device.cu_stream();
+
+    // Kernel outputs
+    let mut n_nonzero = 0;
+    let mut indices: CUdeviceptr = 0;
+
+    let offset = match layout.contiguous_offsets() {
+        Some((start_offset, _)) => start_offset as u64,
+        None => bail!("Input to nonzero kernel op must be contiguous"),
+    };
+
+    unsafe {
+        match storage.dtype() {
+            DType::BF16 => ffi::nonzero_bf16(
+                n_elem,
+                n_dim,
+                shape.as_ptr(),
+                (storage.as_cuda_slice::<bf16>().unwrap().device_ptr() + offset) as *const bf16,
+                &mut indices,
+                &mut n_nonzero,
+                *stream,
+            ),
+            DType::F16 => ffi::nonzero_f16(
+                n_elem,
+                n_dim,
+                shape.as_ptr(),
+                (storage.as_cuda_slice::<f16>().unwrap().device_ptr() + offset) as *const f16,
+                &mut indices,
+                &mut n_nonzero,
+                *stream,
+            ),
+            DType::F32 => ffi::nonzero_f32(
+                n_elem,
+                n_dim,
+                shape.as_ptr(),
+                (storage.as_cuda_slice::<f32>().unwrap().device_ptr() + offset) as *const f32,
+                &mut indices,
+                &mut n_nonzero,
+                *stream,
+            ),
+            DType::F64 => ffi::nonzero_f64(
+                n_elem,
+                n_dim,
+                shape.as_ptr(),
+                (storage.as_cuda_slice::<f64>().unwrap().device_ptr() + offset) as *const f64,
+                &mut indices,
+                &mut n_nonzero,
+                *stream,
+            ),
+            DType::U8 => ffi::nonzero_u8(
+                n_elem,
+                n_dim,
+                shape.as_ptr(),
+                (storage.as_cuda_slice::<u8>().unwrap().device_ptr() + offset) as *const u8,
+                &mut indices,
+                &mut n_nonzero,
+                *stream,
+            ),
+            DType::U32 => ffi::nonzero_u32(
+                n_elem,
+                n_dim,
+                shape.as_ptr(),
+                (storage.as_cuda_slice::<u32>().unwrap().device_ptr() + offset) as *const u32,
+                &mut indices,
+                &mut n_nonzero,
+                *stream,
+            ),
+            DType::I64 => ffi::nonzero_i64(
+                n_elem,
+                n_dim,
+                shape.as_ptr(),
+                (storage.as_cuda_slice::<i64>().unwrap().device_ptr() + offset) as *const i64,
+                &mut indices,
+                &mut n_nonzero,
+                *stream,
+            ),
+        }
+    };
+
+    let output_slice = unsafe {
+        storage
+            .device()
+            .upgrade_device_ptr::<i64>(indices, n_nonzero * n_dim)
```

Comment:
I can't seem to find the implementation code on my end, but is the `i64` type parameter intentional? Surprising, since we're dealing with a pointer!

Reply:
It is. The `CudaSlice` type that this converts to is typed by element (`CudaSlice<T>`) and we have to tell the compiler what the type of `T` is.

https://docs.rs/cudarc/latest/cudarc/driver/safe/struct.CudaDevice.html#method.upgrade_device_ptr